src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs

```diff
@@ -257,6 +258,88 @@ internal string GetString(int index, JsonTokenType expectedType)
             return lastString;
         }
 
+        internal bool TextEquals(int index, ReadOnlySpan<char> otherText, bool isPropertyName)
+        {
+            CheckNotDisposed();
+
+            int matchIndex = isPropertyName ? index - DbRow.Size : index;
+
+            (int lastIdx, string lastString) = _lastIndexAndString;
+        
+            if (lastIdx == matchIndex)
+            {
+                return otherText.SequenceEqual(lastString);
+            }
+
+            byte[] otherUtf8TextArray = null;
+
+            int length = checked(otherText.Length * JsonConstants.MaxExpansionFactorWhileTranscoding);
+            Span<byte> otherUtf8Text = length <= JsonConstants.StackallocThreshold ?
+                stackalloc byte[length] :
+                (otherUtf8TextArray = ArrayPool<byte>.Shared.Rent(length));
+
+            ReadOnlySpan<byte> utf16Text = MemoryMarshal.AsBytes(otherText);
+            OperationStatus status = JsonWriterHelper.ToUtf8(utf16Text, otherUtf8Text, out int consumed, out int written);
+            Debug.Assert(status != OperationStatus.DestinationTooSmall);
+            if (status > OperationStatus.DestinationTooSmall)   // Equivalent to: (status == NeedMoreData || status == InvalidData)
+            {
+                return false;
```

Comment:
What do you mean by pool return here?

Reply:
Good point, I grabbed that code from Utf8JsonReader so that would have the same problem.

Reply:
This can be fixed in both places, separately, imo. I want to think about whether it is absolutely necessary to do so.

Reply:
Sure. I can make changes for both in a separate PR. 
It will look like this: https://github.com/dotnet/corefx/commit/c44726b21f631079cc64a258883371a4411fc0e5

Reply:
Since this isn't really an error/exception case (and we just return false for the comparison), I do think we should return to the arraypool in both cases.

> I can make changes for both in a separate PR. It will look like this

Looks good to me.

Reply:
This is left to do.